### ISSUE TITLE
Pass topic section name to filter function

### ISSUE THREAD

**Issue by robdejonge** [jpmens/mqttwarn#327]
**Current behavior**
When calling a custom function from the `filter` property in a topic section of the INI file, currently the `topic` and `payload` get passed on to the filter function.

**Desired behavior**
In addition to the above, pass the name of the topic section (the bit between the square brackets) on to the custom function called too.

**Comment by robdejonge** [comment]
I can't mark the above as an enhancement request. Can someone do this for me, so that it shows up in the correct queue.

**Comment by robdejonge** [comment]
Further explanation

I wrote a custom filter function, which decides based on the contents of a JSON file if it should block a notification or not : 

```python
def togglestate(topic, message): 

    filename = "warntoggle.json"
    default_topicblock = False

    try:
        with open(filename) as infile:    
            toggles = json.load(infile)
        infile.close()

        if topic in toggles: 
            # file found, topic found
            topicblock = toggles[topic]
        else:
            # file found, adding new topic
            toggles[topic] = default_topicblock
            with open(filename, 'w') as outfile: 
                json.dump(toggles, outfile)
            outfile.close()
            topicblock = False

    except: 
        # file not found or other error
        topicblock = default_topicblock 

    return topicblock

```

And an accompanying web-based interface that lets me edit the content of that JSON file : 

```python
#!/usr/bin/python

import json

def index(req):

    writetoggles = {}
    readtoggles = {} 
    filename = "/var/www/html/warntoggle/warntoggle.json"

    html  = "<html><body><h1>mqttwarn notification toggles</h1>"


    # write data to the file if form was submitted

    if bool(req.form): 

        for k,v in req.form.items():
            writetoggles[k] = eval(str(v))

        with open(filename, 'w') as outfile: 
            json.dump(writetoggles, outfile)
        outfile.close()


    # read data from the file

    with open(filename, 'r') as data_file:
              readtoggles = json.load(data_file)
    data_file.close()


    # display the form

    html += "<table>"
    html += "<form action=\"warntoggle.py\" method=\"post\">"

    for key, value in readtoggles.iteritems():
        if value == True:
            block_checked = " checked"
            notify_checked = ""
        else:
            block_checked = ""
            notify_checked = " checked"

        html += "<tr><td>" + key + "</td><td><input type='radio' name='" + key + "' value='True'" + block_checked + ">Block <input type='radio' name='" + key + "' value='False'" + notify_checked+ ">Notify</td></tr>"

    html += "</table><br/>"
    html += "<nobr><input type='submit' value='Submit'> <a href='warntoggle.py'>reload</a></nobr>"
    html += "</form>"

    html += "</body></html>"

    return html 
```

As the custom function adds unknown topics into the JSON file, I can just add the function to any topic section in `mqttwarn.ini` and it will (as above) let notifications through initially but let me turn them off easily. 

Filtering per topic is what is possible given the current setup, but it would be useful if I could do this by topic section. Hence the enhancement request.

**Comment by rgitzel** [comment]
@robdejonge I have taken a shot at it.  If you are able to run mqttwarn from the command line, clone my repo and change to the 'pass-section' branch:  https://github.com/rgitzel/mqttwarn/pull/2

Then change your filter functions to take a third parameter, which will be the section name.

I did a quick test, and the section name (even just '#') is passed through correctly.

**But there's a problem:**  this will break on all existing two-parameter filter functions, as the number of parameters do not match.

Can you do some testing with your own (three-parameter) filters, to verify the behaviour is what you are looking for?  If so, then we can figure a way around it.

**Comment by robdejonge** [comment]
Tried it and it works perfectly. In the custom function above, added the third parameter for the topic section name and replaced `topic` inside the function for the third parameter, and it all worked well.

To address the problem you raise, I guess it would be possible to introduce a new property for use in the INI file. Just like `alldata` can do the same and more as `datamap`, so could `allfilter` (?) do the same and more as `filter`. It's either this, or a cleanup of the code and deprecation of older functions over time if this is what is preferred. I wouldn't mind either.

If possible, it would be great to also have the `srv` object passed on to the filter function as well. This would allow functions access to the mqttwarn logging object, so I can write there rather than having to write to the system log for debugging purposes. This quite frankly would be useful for all places where custom functions are allowed. 

Thanks so much for working this out in the main code. I'm not sure how @jpmens feels about pulling this into the main repository?

**Comment by rgitzel** [comment]
@robdejonge I found the function to determine the number of arguments, so that's not a problem. I won't claim it's the ideal way, but it's possible. The PR has the change.

Logging from filter functions seems possible already, at least somewhat:  when I `import logging` and then later `logging.debug("foo")` in my filter function then `foo` shows up in the `mqttwarn.log` file with all the usual timestamping and such.  I'm not sure if it has the same configuration as the main program, but I suspect it must, if it's written to that file.

**Comment by jpmens** [comment]
Nice work.

We can take this under the premise that existing `filter`s don't break. So, it's either add a parameter to the end of filter() or create a new name. I tend to prefer the former, in particular as @rgitzel has found out how to do so, but I think the signature would then have to be

```python
def filter(self, name, topic, payload, section=None):
```

i.e. put `section` at end.

And yes, while you're at it, see if you can bring in the `srv` object in order to cater for existing MQTT connections and logging; that would then, I think, finalize it.

**Comment by rgitzel** [comment]
@jpmens I hadn't looked around at the other function functions until now... definitely should pass it in `srv` to be consistent.  

I see you've handled a similar case of varying parameters already:  https://github.com/jpmens/mqttwarn/blob/master/mqttwarn.py#L202   That's a simpler way, and faster I'm sure.

I also see you've more-recently used named parameter calls, is that preferred?

@robdejonge  I've updated the PR. You can use `srv` now.

**Comment by jpmens** [comment]
I don't have a particular preference, but @amotl knows a lot more about this than I do. If he doesn't complain, go ahead as you please. :-)

**Comment by rgitzel** [comment]
I've created a proper PR, including updated documentation.

I'll create a similar PR for the `development` branch.

**Comment by amotl** [comment]
Hi there,

i am **very happy** to see that many people are chiming in to take care about advancing _mqttwarn_'s internals, especially the interfaces between the different subsystems. Welcome, @rgitzel and @robdejonge!

I will try to support that work by providing the promised test harness around _mqttwarn_ to finally get us started with having unit- and integration tests in place. I will also try to add the missing bits of #127 to be able to merge into `master` and finally start releasing packages.

When actually getting back to work i will start by rebasing the `develop` branch onto the current `master`, followed by a force-push. So please be aware of that when pulling the `develop` branch next time - you might have to reset your working tree accordingly. ¹

Cheers,
Andreas.

¹ Please don't hesitate to share your objections. @rgitzel: As i can see #329 and #330 (thanks for that!) are not merged yet, i might try to take this into account as well, at least on behalf of the `develop` branch by applying #330. @jpmens: Should we just merge both PRs?

**Comment by rgitzel** [comment]
@amotl Glad to be of help! :-)

Hold off on those two PRs for the moment, there seems to be a bug.

**Comment by jpmens** [comment]
@amotl glad you're alive! :-)

Give the guys a few moments to solve an issue which is apparently pending. I'm on it and as soon as I hear from @rgitzel I will PRESS THE BUTTONS.

**Comment by robdejonge** [comment]
Closed with https://github.com/jpmens/mqttwarn/pull/329

**Comment by jpmens** [comment]
and doubly-closed via #330 :-)
